### PR TITLE
Action abstraction

### PR DESCRIPTION
**Pull request by davkhech**

### CODE REVIEW

File: test/action_abstraction_tests.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                                             action_len=param['action_len'], 
+                                             available_money=param['available_money'], 
+                                             minimum_bet=param['minimum_bet'], 
+                                             pot=param['pot'], 
+                                             stochastic=param['stochastic'])
+
+                self.assertEqual(math.floor(res*100)/100, param['label_2'])
+
+
+if __name__ == '__main__':
```
Comment (davkhech) [inline test/action_abstraction_tests.py:76]: @MelsHakobyan96 remove this

File: test/action_abstraction_tests.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                                         minimum_bet=param['minimum_bet'],
+                                         increment=param['increment'], 
+                                         stochastic=param['stochastic'])
+
+                self.assertEqual(res, param['label_0'])
+
+
+
+
+
```
Comment (davkhech) [inline test/action_abstraction_tests.py:43]: @MelsHakobyan96 remove these unnecessary blank lines, try to work in scope of PEP8

File: test/action_abstraction_tests.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                                               action_len=param['action_len'],
+                                               available_money=param['available_money'], 
+                                               minimum_bet=param['minimum_bet'], 
+                                               stochastic=param['stochastic'])
+
+                self.assertEqual(res, param['label_1'])
+
+
+
+
```
Comment (davkhech) [inline test/action_abstraction_tests.py:60]: here also

File: test/action_abstraction_tests.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
+import unittest
```
Comment (davkhech) [inline test/action_abstraction_tests.py:1]: move this file under `test/cfr_test/action_abstraction.py`

File: src/cfr/action_abstraction.py [hunk 5]
```diff
@@ -77,53 +80,56 @@ def action_abstraction_count(bet, action_len, available_money, minimum_bet, stoc
...
+        
+    for j, i in enumerate(upscaled_actions[:-1]):
+        if i < bet and upscaled_actions[j+1] > bet:
+            B, A = upscaled_actions[j+1], i
+            
+    upscaled_action = action_abstraction(bet, available_money=available_money, minimum_bet=minimum_bet, A=A, B=B, stochastic=stochastic)
+    action_index = upscaled_actions.index(upscaled_action)
+    action = actions[action_index]
+        
+    return action
```
Comment (davkhech) [inline src/cfr/action_abstraction.py:135]: let there be just a single newline at the end of file

File: src/cfr/action_abstraction.py [hunk 6]
```diff
@@ -77,53 +80,56 @@ def action_abstraction_count(bet, action_len, available_money, minimum_bet, stoc
...
 
     actions = list(range(minimum_bet, available_money, increment))
+    if not bool(actions):
+        actions = [minimum_bet, available_money]
+
     actions[-1] = available_money
 
     for j, i in enumerate(actions[:-1]):
         if i < bet and actions[j+1] > bet:
             B, A = actions[j+1], i
```
Comment (davkhech) [inline src/cfr/action_abstraction.py:92]: remove unnecessary empty lines

File: src/cfr/action_abstraction.py [hunk 7]
```diff
@@ -1,71 +1,74 @@
...
+        B_prob = 1 - A_prob
 
-	else:
+    else:
 
-		median = int(round((2*A*B + (A + B)) / (A + B + 2)))
+        median = int(round((2*A*B + (A + B)) / (A + B + 2)))
 
-		if bet > median:
+        if bet > median:
```
Comment (davkhech) [inline src/cfr/action_abstraction.py:47]: remove

File: test/action_abstraction_tests.py (hunk 1 above)
Comment (MelsHakobyan96) [reply test/action_abstraction_tests.py:76]: aranc dra chi ashxatum

Comment (davkhech) [reply test/action_abstraction_tests.py:76]: it works
no need to run the test by itself, you need to run it through terminal. for example
`python -m nose test/cft_test/action_abstraction.py`

Comment (davkhech) [reply test/action_abstraction_tests.py:76]: when you put the test into it's right place run our custom script
`./run_tests.sh` and see if all the tests pass

File: src/cfr/action_abstraction.py (hunk 7 above)
Comment (MelsHakobyan96) [reply src/cfr/action_abstraction.py:47]: WHY????

File: test/action_abstraction_tests.py (hunk 1 above)
Comment (MelsHakobyan96) [reply test/action_abstraction_tests.py:76]: haaa lav haskaca

File: test/action_abstraction_tests.py (hunk 2 above)
Comment (MelsHakobyan96) [reply test/action_abstraction_tests.py:43]: done

File: test/action_abstraction_tests.py (hunk 3 above)
Comment (MelsHakobyan96) [reply test/action_abstraction_tests.py:60]: done

File: test/action_abstraction_tests.py (hunk 4 above)
Comment (MelsHakobyan96) [reply test/action_abstraction_tests.py:1]: done

File: src/cfr/action_abstraction.py (hunk 5 above)
Comment (MelsHakobyan96) [reply src/cfr/action_abstraction.py:135]: done

File: src/cfr/action_abstraction.py (hunk 6 above)
Comment (MelsHakobyan96) [reply src/cfr/action_abstraction.py:92]: done

File: src/cfr/action_abstraction.py (hunk 7 above)
Comment (davkhech) [reply src/cfr/action_abstraction.py:47]: I mean the empty lines :D

Comment (davkhech) [reply src/cfr/action_abstraction.py:47]:
``` 
if smth:

    do smth:
```

Comment (MelsHakobyan96) [reply src/cfr/action_abstraction.py:47]: done

**Pull request merged**